# WINE 분류해보기 실습

- 필요한 모듈 Import 하기

In [31]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

- 데이터 준비하기

In [32]:
wines = load_wine()
wines_data = wines.data
wines_label = wines.target

 - 데이터 이해하기
 1. feature data 지정하기
 2. label data 지정하기
 3. Target Names 출력해보기
 4. 데이터 Describe 해보기

In [33]:
print("\n데이터의 키값 \n", wines.keys())

print("\n데이터 모양 \n","데이터 갯수 : ",wines.data.shape[0],"\n데이터 크기 : ", wines.data.shape[1])

print("데이터 셈플\n", wines_data[0])

print(wines.feature_names)
print("\n 타켓 이름 : \n", wines.target_names)


데이터의 키값 
 dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

데이터 모양 
 데이터 갯수 :  178 
데이터 크기 :  13
데이터 셈플
 [1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
 2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']

 타켓 이름 : 
 ['class_0' 'class_1' 'class_2']


In [52]:
import pandas as pd

print("데이터 값 : \n", wines.data[0])
df_wines= pd.DataFrame(data = wines_data, columns=wines.feature_names)

df_wines.describe()

데이터 값 : 
 [1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
 2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]


alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  
count       178.000000  178.000000                    178.000000   178.000000  
mean          5.058090    0.957449                      2.611685   746.893258  
std           2.318286    0.228572                      0.709990   314.907474  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.220000    0.782500                      1.937500   500.500000  
50%           4.690000    0.965000                      2.780000   673.500000  
75%           6.200000    1.120000                      3.170000   985.000000  
max          13.000000    1.710000                      4.000000  1680.000000

In [35]:
print(wines_label)
print(wines.target_names)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]
['class_0' 'class_1' 'class_2']


 - 데이터셋 분류하기

In [36]:
X_train,X_test,y_train,y_test=train_test_split(wines_data,
                                               wines_label,
                                               test_size=0.2,
                                               random_state=70)
from sklearn import preprocessing
import numpy as np
X_train = preprocessing.scale(X_train)
X_test = preprocessing.scale(X_test)

 - 모델 불러오기

In [37]:
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)

from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)

from sklearn import svm
svm_model = svm.SVC()

from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

accuracy_model={}

In [38]:
#Decision Tree
decision_tree.fit(X_train, y_train)
decision_y_predict=decision_tree.predict(X_test)
char_Decision=classification_report(y_test,decision_y_predict)
print(char_Decision)

              precision    recall  f1-score   support

           0       0.85      0.85      0.85        13
           1       0.67      0.73      0.70        11
           2       0.91      0.83      0.87        12

    accuracy                           0.81        36
   macro avg       0.81      0.80      0.80        36
weighted avg       0.81      0.81      0.81        36



In [39]:
#RandomForest
random_forest.fit(X_train, y_train)
random_forest_y_predict=random_forest.predict(X_test)
char_Random=classification_report(y_test,random_forest_y_predict)
print(char_Random)

              precision    recall  f1-score   support

           0       1.00      0.92      0.96        13
           1       0.85      1.00      0.92        11
           2       1.00      0.92      0.96        12

    accuracy                           0.94        36
   macro avg       0.95      0.95      0.94        36
weighted avg       0.95      0.94      0.95        36



In [40]:
#SVM
svm_model.fit(X_train, y_train)
svm_model_y_predict=svm_model.predict(X_test)
char_SVM=classification_report(y_test,decision_y_predict)
print(char_SVM)

              precision    recall  f1-score   support

           0       0.85      0.85      0.85        13
           1       0.67      0.73      0.70        11
           2       0.91      0.83      0.87        12

    accuracy                           0.81        36
   macro avg       0.81      0.80      0.80        36
weighted avg       0.81      0.81      0.81        36



In [41]:
#SGD
sgd_model.fit(X_train, y_train)
sgd_model_y_predict=sgd_model.predict(X_test)
char_SGD=classification_report(y_test,sgd_model_y_predict)
print(classification_report(y_test,sgd_model_y_predict))

              precision    recall  f1-score   support

           0       0.92      0.92      0.92        13
           1       0.71      0.91      0.80        11
           2       1.00      0.75      0.86        12

    accuracy                           0.86        36
   macro avg       0.88      0.86      0.86        36
weighted avg       0.88      0.86      0.86        36



In [42]:
#logistic Regression
logistic_model.fit(X_train, y_train)
logistic_model_y_predict=logistic_model.predict(X_test)
char_Logistic=classification_report(y_test,sgd_model_y_predict)
print(char_Logistic)

              precision    recall  f1-score   support

           0       0.92      0.92      0.92        13
           1       0.71      0.91      0.80        11
           2       1.00      0.75      0.86        12

    accuracy                           0.86        36
   macro avg       0.88      0.86      0.86        36
weighted avg       0.88      0.86      0.86        36



In [43]:

print(accuracy_model)

{}


## WINE 데이터셋 학습 결과
## ACCURACY비교 결과, RANDOM FOREST의 정확도가 94% 로 가장 좋다.